rust/connlib/tunnel/src/tests/transition.rs

```diff
@@ -107,78 +103,163 @@ pub(crate) enum DnsTransport {
+pub(crate) struct Seq(pub u16);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct Identifier(pub u16);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct SPort(pub u16);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub(crate) struct DPort(pub u16);
+
+#[derive(Debug, Clone)]
+#[expect(clippy::large_enum_variant)]
+pub(crate) enum Destination {
+    DomainName {
+        resolved_ip: sample::Selector,
+        name: DomainName,
+    },
+    IpAddr(IpAddr),
+}
+
+/// Helper enum
+#[derive(Debug, Clone)]
+enum PacketDestination {
+    DomainName(DomainName),
+    IpAddr(IpAddr),
+}
+
+pub(crate) trait ReplyTo {
+    fn reply_to(self) -> Self;
+}
+
+impl ReplyTo for (SPort, DPort) {
+    fn reply_to(self) -> Self {
+        (SPort(self.1 .0), DPort(self.0 .0))
+    }
+}
+
+impl ReplyTo for (Seq, Identifier) {
+    fn reply_to(self) -> Self {
+        self
+    }
+}
+
+impl From<DomainName> for PacketDestination {
+    fn from(name: DomainName) -> Self {
+        PacketDestination::DomainName(name)
+    }
+}
+
+impl From<Ipv4Addr> for PacketDestination {
+    fn from(addr: Ipv4Addr) -> Self {
+        PacketDestination::IpAddr(addr.into())
+    }
+}
+
+impl From<Ipv6Addr> for PacketDestination {
+    fn from(addr: Ipv6Addr) -> Self {
+        PacketDestination::IpAddr(addr.into())
+    }
+}
+
+impl From<IpAddr> for PacketDestination {
+    fn from(addr: IpAddr) -> Self {
+        PacketDestination::IpAddr(addr)
+    }
+}
+
+impl PacketDestination {
+    fn into_destination(self, resolved_ip: sample::Selector) -> Destination {
+        match self {
+            PacketDestination::DomainName(name) => Destination::DomainName { resolved_ip, name },
+            PacketDestination::IpAddr(addr) => Destination::IpAddr(addr),
+        }
+    }
+}
+
+#[expect(private_bounds)]
+pub(crate) fn icmp_to_destination<I, D>(
     src: impl Strategy<Value = I>,
-    dst: impl Strategy<Value = I>,
+    dst: impl Strategy<Value = D>,
 ) -> impl Strategy<Value = Transition>
 where
     I: Into<IpAddr>,
+    D: Into<PacketDestination>,
 {
     (
         src.prop_map(Into::into),
         dst.prop_map(Into::into),
         any::<u16>(),
         any::<u16>(),
+        any::<sample::Selector>(),
         any::<u64>(),
     )
-        .prop_map(|(src, dst, seq, identifier, payload)| {
-            Transition::SendICMPPacketToNonResourceIp {
+        .prop_map(|(src, dst, seq, identifier, resolved_ip, payload)| {
+            Transition::SendIcmpPacket {
                 src,
-                dst,
-                seq,
-                identifier,
+                dst: dst.into_destination(resolved_ip),
+                seq: Seq(seq),
+                identifier: Identifier(identifier),
                 payload,
             }
         })
 }
 
-pub(crate) fn icmp_to_cidr_resource<I>(
+#[allow(private_bounds)]
```

Comment:
```suggestion
#[expect(private_bounds)]
```

We shouldn't use `allow` because it doesn't fail when it is no longer necessary.